x/cdp/keeper/fees.go

```diff
@@ -27,44 +27,52 @@ func (k Keeper) CalculateFees(ctx sdk.Context, principal sdk.Coins, periods sdk.
 
-	liquidationRatio := k.getLiquidationRatio(ctx, collateralDenom)
-	priceDivLiqRatio := price.Price.Quo(liquidationRatio)
-	if priceDivLiqRatio.IsZero() {
-		priceDivLiqRatio = sdk.SmallestDec()
-	}
-	// NOTE - we have a fixed cutoff at 110% - this may or may not be changed in the future
-	normalizedRatio := sdk.OneDec().Quo(priceDivLiqRatio).Mul(sdk.MustNewDecFromStr("1.1"))
+	k.IterateAllCdps(ctx, func(cdp types.CDP) bool {
 
-	// now iterate over all the cdps based on collateral ratio
-	k.IterateCdpsByCollateralRatio(ctx, collateralDenom, normalizedRatio, func(cdp types.CDP) bool {
 		oldCollateralToDebtRatio := k.CalculateCollateralToDebtRatio(ctx, cdp.Collateral, cdp.Principal.Add(cdp.AccumulatedFees...))
-		// get the number of periods
 		periods := sdk.NewInt(ctx.BlockTime().Unix()).Sub(sdk.NewInt(cdp.FeesUpdated.Unix()))
 
-		// now calculate and store additional fees
-		additionalFees := k.CalculateFees(ctx, cdp.Principal, periods, collateralDenom)
+		newFees := k.CalculateFees(ctx, cdp.Principal, periods, collateralDenom)
+
+		if newFees.IsZero() {
+			return false
+		}
+
+		dp, _ := k.GetDebtParam(ctx, principalDenom)
+		savingsRate := dp.SavingsRate
+
+		newFeesSavings := sdk.NewDecFromInt(newFees.AmountOf(principalDenom)).Mul(savingsRate).RoundInt()
+		newFeesSurplus := newFees.AmountOf(principalDenom).Sub(newFeesSavings)
 
-		// now add the additional fees to the accumulated fees for the cdp
-		cdp.AccumulatedFees = cdp.AccumulatedFees.Add(additionalFees...)
+		if newFeesSavings.IsZero() || newFeesSurplus.IsZero() {
+			return false
+		}
+
+		// mint debt coins to the cdp account
+		k.MintDebtCoins(ctx, types.ModuleName, k.GetDebtDenom(ctx), newFees)
+		previousDebt := k.GetTotalPrincipal(ctx, collateralDenom, principalDenom)
+		feeCoins := sdk.NewCoins(sdk.NewCoin(principalDenom, previousDebt))
+		k.SetTotalPrincipal(ctx, collateralDenom, principalDenom, feeCoins.Add(newFees...).AmountOf(principalDenom))
+
+		// mint surplus coins divided between the liquidator and savings module accounts.
+		k.supplyKeeper.MintCoins(ctx, types.LiquidatorMacc, sdk.NewCoins(sdk.NewCoin(principalDenom, newFeesSurplus)))
+		k.supplyKeeper.MintCoins(ctx, types.SavingsRateMacc, sdk.NewCoins(sdk.NewCoin(principalDenom, newFeesSavings)))
+
+		// TODO QUESTION DO WE STILL NEED TO UPDATE THE ACCUMULATED FEES SINCE WE ALREADY ADDED TO THE PRINCIPAL?
+		// now add the new fees fees to the accumulated fees for the cdp
+		cdp.AccumulatedFees = cdp.AccumulatedFees.Add(newFees...)
 
 		// and set the fees updated time to the current block time since we just updated it
 		cdp.FeesUpdated = ctx.BlockTime()
 		collateralToDebtRatio := k.CalculateCollateralToDebtRatio(ctx, cdp.Collateral, cdp.Principal.Add(cdp.AccumulatedFees...))
 		k.RemoveCdpCollateralRatioIndex(ctx, cdp.Collateral[0].Denom, cdp.ID, oldCollateralToDebtRatio)
 		k.SetCdpAndCollateralRatioIndex(ctx, cdp, collateralToDebtRatio)
+
+		k.SetCDP(ctx, cdp)
```

Comment:
Ah, just realized that `SetCdpAndCollateralRatioIndex` already calls `SetCDP`, so we don't need this line.

Reply:
🆗  I will remove this

Reply:
done